drivers/can/can_net.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2019 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <net/can.h>
+#include <net/net_pkt.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_can, CONFIG_CAN_NET_LOG_LEVEL);
+
+#define CAN_NET_FILTER_NOT_SET -1
+
+struct net_can_context {
+	struct device *can_dev;
+	struct net_if *iface;
+	int recv_filter_id;
+	int mcast_filter_id;
+};
+
+static inline u8_t get_frame_datalength(struct zcan_frame *frame)
+{
+	/* TODO: Needs update when CAN FD support is added */
+	return frame->dlc;
+}
+
+static inline u16_t get_can_lladdr_src(struct zcan_frame *frame)
+{
+	return (frame->ext_id >> CAN_NET_IF_ADDR_SRC_POS) &
+		     CAN_NET_IF_ADDR_MASK;
+}
+
+static inline u16_t get_can_lladdr_dest(struct zcan_frame *frame)
+{
+	return (frame->ext_id >> CAN_NET_IF_ADDR_DEST_POS) &
+		     CAN_NET_IF_ADDR_MASK;
+}
+
+static inline void set_lladdr(struct net_pkt *pkt, struct zcan_frame *frame)
+{
+	struct net_buf *buf = pkt->buffer;
+
+	net_pkt_lladdr_dst(pkt)->addr = buf->data;
```

Comment:
>You receive empty pkt?

This pkt was allocated before and is empty, yes.

>The you try to store can_lladdr's at the beginning of it, point other pkt lladdr fields inside its own buffer?

That is correct. Other l2 do the same. ETH, for example, sets this the ptr to the address in the ETH header and then pull it. The problem is that there is no other memmory reserved for lladdr in the pkt. I guess it's because of the different sizes.

>You use different values for the sizes of these field, like when it's referred as and address, it's sizeof(struct net_can_lladdr), while when you put it into buffer, it's sizeof(u16_t)? Do you expect them to match?

`sizeof(struct net_can_lladdr)` is actually 2 octets. I need to change the byte order to net-byte-order, so I have to use u16_t as type.

Reply:
> Other l2 do the same.

Ok. (I assume they do it the same with the same, uncommented code.)

> sizeof(struct net_can_lladdr) is actually 2 octets.

Yeah, I figured. What's somewhat confusing is why in one case you use `sizeof(struct net_can_lladdr)` and in other `sizeof(u16_t)`.

> I need to change the byte order to net-byte-order, so I have to use u16_t as type.

Noting that, it indeed makes sense to use different types there, though it still looks a bit weird at first sight. I'd personally looked to add a code comment if I wrote such code (not sure if I'd succeed).

All in all, that's relatively minor point comparing with the rest of code.